plugins/utilities/Updated_Ultra_Partywindow.py

```diff
@@ -0,0 +1,2768 @@
+
+
+class ColorTracker:
+    def __init__(self):
+        self.saved = {}
+
+    def _get_safe_color(self, sender):
+        while True:
+            color = (random.random(), random.random(), random.random())
+            s = 0
+            background = ba.app.config.get('PartyWindow Main Color', (0.5, 0.5, 0.5))
+            for i, j in zip(color, background):
+                s += (i - j) ** 2
+            if s > 0.1:
+                self.saved[sender] = color
+                if len(self.saved) > 20:
+                    self.saved.pop(list(self.saved.keys())[0])
+                break
+            time.sleep(0.1)
+
+    def _get_sender_color(self, sender):
+        if sender not in self.saved:
+            self.thread = Thread(target=self._get_safe_color, args=(sender,))
+            self.thread.start()
+            return (1, 1, 1)
+        else:
+            return self.saved[sender]
+
+
+class PrivateChatHandler:
+    def __init__(self):
+        self.pvt_msgs = {}
+        self.login_id = None
+        self.last_msg_id = None
+        self.logged_in = False
+        self.cookieProcessor = urllib.request.HTTPCookieProcessor()
+        self.opener = urllib.request.build_opener(self.cookieProcessor)
+        self.filter = 'all'
+        self.pending_messages = []
+        self.friends_status = {}
+        self.error = ''
+        Thread(target=self._ping).start()
+
+    def _load_ids(self):
+        with open(saved_ids_file, 'r') as f:
+            saved = json.load(f)
+            if self.myid in saved:
+                self.saved_ids = saved[self.myid]
+            else:
+                self.saved_ids = {'all': '<all>'}
+
+    def _dump_ids(self):
+        with open(saved_ids_file, 'r') as f:
+            saved = json.load(f)
+        with open(saved_ids_file, 'w') as f:
+            saved[self.myid] = self.saved_ids
+            json.dump(saved, f)
+
+    def _ping(self):
+        self.server_online = False
+        response = self._send_request(url=f'{url}')
+        if not response:
+            self.error = 'Server offline'
+        elif response:
+            try:
+                self.server_online = True
+                version = float(response.replace('v', ''))
+                if version > __version__:
+                    self._update_version()
+            except:
+                self.error = 'Server offline'
+
+    def _update_version(self):
+        new_file = self._send_request(url=f'{url}/updatepartywindow?get=file')
+        hash = self._send_request(url=f'{url}/updatepartywindow?get=md5')
+        if hash and new_file:
+            file_hash = md5(new_file.encode()).hexdigest()
+            if hash == file_hash:
+                with open(__file__, 'w') as f:
+                    f.write(new_file)
+                    _ba.pushcall(ba.Call(ba.screenmessage,
+                                 'Ultra party window updated\nNeeds restart.'), True)
```

Comment:
can we remove this auto self-updating code here as this can now be handled through plugin manager?

Reply:
or let's at least provide an option somewhere if the user would like to disable auto-updates

Reply:
removed auto updating codes :)